File: Students/BinQingZheng/Week_1_HW/TwoPointers/lc16.js
// Problem Statement #
// Given an array of unsorted numbers and a target number, find a triplet in the array whose sum is as close to the target number as possible, return the sum of the triplet. 
// If there are more than one such triplet, return the sum of the triplet with the smallest sum.

const triplet_sum_close_to_target = function (arr, target_sum) {
    arr.sort((a, b) => a-b);
    let best_sum = arr[0] + arr[1] + arr[2];
    for (let i = 0; i<arr.length-2; i++) {
      let low = i+1, high = arr.length-1;
      while (low < high) {
        const sum = arr[i] + arr[low] +arr[high];
        if (sum > target_sum) {
          high -=1;
        } else {
          low +=1;
        }
        if (Math.abs(target_sum - sum) < Math.abs(target_sum - best_sum)) {
          best_sum = sum;
        }
      }
    }
    return best_sum;
};
  
console.log(triplet_sum_close_to_target([-2, 0, 1, 2], 2));
console.log(triplet_sum_close_to_target([-3, -1, 1, 2], 1));
console.log(triplet_sum_close_to_target([1, 0, 1, 1], 100));

//Time Complexity: O(n^2)
//Space Complexity: O(1)

File: Students/Xin_Jin/lc16.java
package BootCamp;

import java.util.Arrays;

public class lc16 {
    public static void main(String[] args) {
        System.out.println(threeSumClosest(new int[]{-2, 0, 1, 2},2));
        System.out.println(threeSumClosest(new int[]{-3, -1, 1, 2},1));
        System.out.println(threeSumClosest(new int[]{-1, 0, 1, 1},100));
    }
    public static int threeSumClosest(int[] nums, int target) {
        int closestNum=Integer.MAX_VALUE;
        Arrays.sort(nums);
        for(int i=0;i<nums.length-2;i++)
        {
            int low=i+1;
            int high=nums.length-1;
            while(low<high)
            {
                int sum=nums[i]+nums[low]+nums[high];
                if(Math.abs(target-sum)<Math.abs(closestNum))
                {
                    closestNum=target-sum;
                }
                if(sum<target)low++;
                else{high--;};
            }
        }
        return target-closestNum;
    }
}
